src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,480 @@
+        {
+            ConcurrentBag<Statistic> sums = new ConcurrentBag<Statistic>();
+
+            Parallel.ForEach(projects.SelectMany(i => i.Documents), document =>
+            {
+                var documentStatistics = GetAnalyzerStatisticsAsync(document, cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult();
+                sums.Add(documentStatistics);
+            });
+
+            Statistic sum = sums.Aggregate(new Statistic(0, 0, 0), (currentResult, value) => currentResult + value);
+            return Task.FromResult(sum);
+        }
+
+        private static async Task<Statistic> GetAnalyzerStatisticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            SyntaxTree tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+
+            SyntaxNode root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var tokensAndNodes = root.DescendantNodesAndTokensAndSelf(descendIntoTrivia: true);
+
+            int numberOfNodes = tokensAndNodes.Count(x => x.IsNode);
+            int numberOfTokens = tokensAndNodes.Count(x => x.IsToken);
+            int numberOfTrivia = root.DescendantTrivia(descendIntoTrivia: true).Count();
+
+            return new Statistic(numberOfNodes, numberOfTokens, numberOfTrivia);
+        }
+
+        private static IEnumerable<DiagnosticAnalyzer> FilterAnalyzers(IEnumerable<DiagnosticAnalyzer> analyzers, Options options)
+        {
+            foreach (var analyzer in analyzers)
+            {
+                if (options.UseAll)
+                {
+                    yield return analyzer;
+                }
+                else if (options.AnalyzerIds.Count == 0)
+                {
+                    if (analyzer.SupportedDiagnostics.Any(i => i.IsEnabledByDefault))
+                    {
+                        yield return analyzer;
+                    }
+
+                    continue;
+                }
+                else if (analyzer.SupportedDiagnostics.Any(y => options.AnalyzerIds.Contains(y.Id)))
+                {
+                    yield return analyzer;
+                }
+            }
+        }
+
+        private static ImmutableArray<Assembly> GetAnalyzerAssemblies(string path)
+        {
+            if (File.Exists(path))
+            {
+                return ImmutableArray.Create(Assembly.LoadFrom(path));
+            }
+            else if (Directory.Exists(path))
+            {
+                var builder = ImmutableArray.CreateBuilder<Assembly>();
+                foreach (var file in Directory.GetFiles(path, "*.dll", SearchOption.AllDirectories))
+                {
+                    builder.Add(Assembly.LoadFrom(file));
+                }
+
+                return builder.ToImmutable();
+            }
+
+            Utilities.WriteLine($"Cannot load assembliy or assemblies from {path}.", ConsoleColor.Red);
```

Comment:
so this only takes care of the case where file/directory path doesn't exist, but doesn't catch exceptions for assembly load.

Reply:
Now, we'd use
 var analyzerReference = new AnalyzerFileReference(path, AssemblyLoader.Instance);
            return analyzerReference.GetAnalyzers(LanguageNames.CSharp);

if this fails, the whole application fails and display the exception. I'm not sure we need to support partial loads.

---
In reply to: [149861847](https://github.com/dotnet/roslyn/pull/23087#discussion_r149861847) [](ancestors = 149861847)